### PR TITLE
update master

### PR DESCRIPTION
**Pull request by nponarul**

### CODE REVIEW

File: bower.json [hunk 1]
```diff
@@ -62,5 +62,8 @@
         "build/jsfeat.js"
       ]
     }
+  },
+  "resolutions": {
```
Comment (alxndrkalinin) [inline bower.json]: Why is this in `<resolutions>` section and not in `<dependencies>`? Were there problems during installation?

Comment (nponarul) [inline bower.json]: They're were problems with the installation of jsfeat. I couldn't figure
out why so I just ignored it. Does it tell you what the error is? I could
try reinstalling and send you the error message

On Wed, Jan 13, 2016 at 2:52 PM, Alexandr Kalinin <notifications@github.com>
wrote:

> In bower.json
> <https://github.com/SOCR/SOCRAT/pull/52#discussion_r49664015>:
>
> > @@ -62,5 +62,8 @@
> >          "build/jsfeat.js"
> >        ]
> >      }
> > +  },
> > +  "resolutions": {
>
> Why is this in <resolutions> section and not in <dependencies>? Were
> there problems during installation?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SOCR/SOCRAT/pull/52/files#r49664015>.
>



-- 
Nellie Ponarul
B.S. in Applied Mathematics
University of California, Los Angeles | 2017
nellie.ponarul@ucla.edu

File: app/partials/analysis/charts/sidebar.jade [hunk 2]
```diff
@@ -1,17 +1,16 @@
...
-            legend Variables
-            label X&nbsp;
-            select(ng-model="xVariable")
-                option(ng-repeat="v in variables") {{v}}
-            label Y&nbsp;
-            select(ng-model="yVariable")
-                option(ng-repeat="v in variables") {{v}}
+div(ng-controller="sidechartsCtrl")
+  h4 Graphs
+  select.form-control#chooseGraph(value="Choose a graph" ng-change="changeName()" ng-model = "graphSelect" ng-options = "g as g.name for g in graphs")
```
Comment (alxndrkalinin) [inline app/partials/analysis/charts/sidebar.jade]: (style): unify usage of spaces: they should always be present after comma, but not around equality sign when you pass parameter/attribute values: i.e. `<ng-model="graphSelect">`, etc.

File: bower.json (hunk 1 above)
Comment (alxndrkalinin) [inline bower.json]: Yea, let's do that - I'll try to repoduce. But `<dcjs>` should be moved to `<dependencies>` if it doesn't break anything.

File: app/partials/analysis/charts/tab-head.jade [hunk 3]
```diff
@@ -0,0 +1,8 @@
+
+
+//Top of graph tabs
```
Comment (alxndrkalinin) [inline app/partials/analysis/charts/tab-head.jade]: (style): please always put a space before comment text, it helps with readability

File: app/scripts/analysis/charts/charts.coffee [hunk 4]
```diff
@@ -1,86 +1,671 @@
...
+    _getMsgList = () ->
+      _msgList
 
-]
+    getSb: _getSb
+    setSb: _setSb
+    getMsgList: _getMsgList
+])
+
+.controller('mainchartsCtrl', [
```
Comment (alxndrkalinin) [inline app/scripts/analysis/charts/charts.coffee]: (style): controller names should be in camelCase and start with module name, i.e. `chartsMainCtrl`

File: app/scripts/analysis/charts/charts.coffee [hunk 5]
```diff
@@ -1,86 +1,671 @@
...
+  (ctrlMngr,$scope) ->
+    console.log 'mainchartsCtrl executed'
+    _chart_data = null
+
+    _updateData = () ->
+      $scope.chartData = _chart_data
+
+    $scope.$on 'charts:graphDiv', (event, data) ->
+      _chart_data = data
+      console.log data
```
Comment (alxndrkalinin) [inline app/scripts/analysis/charts/charts.coffee]: (refactror): remove all logs to console, especially data dumps, unless it's important diagnostic (not debugging) information; best practice is to prepend message with module name, e.g. [see this](https://github.com/SOCR/SOCRAT/blob/master/app/scripts/core/core.coffee#L105)

File: app/scripts/analysis/charts/charts.coffee [hunk 6]
```diff
@@ -1,86 +1,671 @@
...
+  '$rootScope'
+  '$stateParams'
+  '$q'
+  'app_analysis_charts_dataTransform'
+  (ctrlMngr, $scope, $rootScope, $stateParams, $q, dataTransform) ->
+    console.log 'sidechartsCtrl executed'
+    _chartData = null
+    _headers = null
+
+    $scope.selector1={}
```
Comment (alxndrkalinin) [inline app/scripts/analysis/charts/charts.coffee]: (style): in code spaces should be around `=` unless it's a value of parameter that is being passed to a function

File: app/scripts/analysis/charts/charts.coffee [hunk 7]
```diff
@@ -1,86 +1,671 @@
...
+    $scope.selector2={}
+    $scope.selector3={}
+    $scope.graphInfo = {graph:"", x:"", y:"",z:""}
+    $scope.graphs = [{name:'Bar Graph', value:0},{name:'Scatter Plot', value:1},{name:'Histogram', value:2},{name:'Bubble Chart', value:3},{name:'Pie Chart', value:4}]
```
Comment (alxndrkalinin) [inline app/scripts/analysis/charts/charts.coffee]: (style): array of objects should be defined in CoffeeScript style for better readability, e.g. see [app.coffee](https://github.com/nponarul/SOCRAT/blob/charts/app/app.coffee#L149); also, there should be space after semicolon, but not before.

File: app/scripts/analysis/charts/charts.coffee [hunk 8]
```diff
@@ -1,86 +1,671 @@
...
+
+
+
+    $scope.createGraph = () ->
+      graphFormat = () ->
+        obj = []
+        if $scope.graphInfo.y is "" and $scope.graphInfo.z is ""
+          obj = []
+          i=1
+          while i < _chartData[0].length
```
Comment (alxndrkalinin) [inline app/scripts/analysis/charts/charts.coffee]: (refactor): `while` loops should be avoided when possible and `for` loop should be used instead; indices of elements can be accessed like this: `for element, index in array ...'

File: app/scripts/analysis/charts/charts.coffee [hunk 9]
```diff
@@ -1,86 +1,671 @@
...
+            tmp = {}
+            tmp = JSON.stringify({x:_chartData[$scope.graphInfo.x][i].value})
```
Comment (alxndrkalinin) [inline app/scripts/analysis/charts/charts.coffee]: (refactor): not clear why do you use `stringify` and then `parse` here, why not just create an object? also, no need to create empty object first

File: app/scripts/analysis/charts/charts.coffee [hunk 10]
```diff
@@ -1,86 +1,671 @@
...
+            tmp = {}
+            tmp = JSON.stringify({x:_chartData[$scope.graphInfo.x][i].value , y:_chartData[$scope.graphInfo.y][i].value, z:_chartData[$scope.graphInfo.z][i].value})
+            obj.push(JSON.parse(tmp))
+            i++
+
+        console.log obj
+        return obj
+      console.log $scope.graphInfo
+      send = graphFormat()
+      results = {
```
Comment (alxndrkalinin) [inline app/scripts/analysis/charts/charts.coffee]: (style): curly braces are not required here

File: app/scripts/analysis/charts/charts.coffee [hunk 11]
```diff
@@ -1,86 +1,671 @@
...
+      xAxis = null
+      yAxis = null
+      svg = null
+      data = null
+      _graph = null
+      container = null
+      pieData = null
+      gdata = null
+
+      makepieData = (data) ->
```
Comment (alxndrkalinin) [inline app/scripts/analysis/charts/charts.coffee]: (style): function names should be in camelCase (i.e. makePieData)

File: app/scripts/analysis/charts/charts.coffee [hunk 12]
```diff
@@ -1,86 +1,671 @@
...
+        pieMax = d3.max(data, (d)->parseFloat d.x)
+        pieMin = d3.min(data, (d)->parseFloat d.x)
+        a = 0
+        b = 0
+        c = 0
+        d = 0
+        e = 0
+        f = 0
+        g = 0
+        rangeInt = (pieMax - pieMin)/7
```
Comment (alxndrkalinin) [inline app/scripts/analysis/charts/charts.coffee]: (refactor): it's good practice to initialize variables with such 'magic constants' like 7 here in the beginning of function and give them meaningful names

File: app/scripts/analysis/charts/charts.coffee [hunk 13]
```diff
@@ -1,86 +1,671 @@
...
+#        console.log pieMin+rangeInt
+        for da in data
+          val = parseFloat da.x
+
+#          console.log val, pieMin+rangeInt
+
+          if val < (pieMin+rangeInt)
```
Comment (alxndrkalinin) [inline app/scripts/analysis/charts/charts.coffee]: (refactor): if I correctly understand what are you doing here, this is not very elegant implementation of search and sorting of values, you can do something like this:
```
maxPiePieces = 7  # set magic constant to variable
xPoints = data.map(el) -> el.x
piePieces = [0..maxPiePieces - 1]  # create array with numbers of pie pieces
pieceVolumes = pieces.map(pieceNum) ->
  currPiecePoints = xPoints.filter(x) -> pieMin + pieceNum * rangeInt <= x < pieMin + (pieceNum + 1) * rangeInt  # filter data points by limits of each piece
  currPiecePoints.length  # return number of elements in each piece
```

File: app/scripts/analysis/charts/charts.coffee [hunk 14]
```diff
@@ -1,86 +1,671 @@
...
+            c++
+          else if (pieMin+3*rangeInt) <= val < (pieMin+4*rangeInt)
+            d++
+          else if (pieMin+4*rangeInt) <= val < (pieMin+5*rangeInt)
+            e++
+          else if (pieMin+5*rangeInt) <= val < (pieMin+6*rangeInt)
+            f++
+          else if (pieMin+6*rangeInt) <= val < (pieMin+7*rangeInt)
+            g++
+        first = (pieMin+rangeInt).toFixed(2)+"-"+(pieMin).toFixed(2)
```
Comment (alxndrkalinin) [inline app/scripts/analysis/charts/charts.coffee]: (style): add spaces please

File: app/scripts/analysis/charts/charts.coffee [hunk 15]
```diff
@@ -1,86 +1,671 @@
...
+
+
+        x = d3.scale.linear().domain([xMin,xMax]).range([ 0, width ])
+        y = d3.scale.linear().domain([yMin,yMax]).range([ height, 0 ])
+        xAxis = d3.svg.axis().scale(x).orient('bottom')
+        yAxis = d3.svg.axis().scale(y).orient('left')
+
+
+        # values
+        xValue = (d)->parseFloat d.x
```
Comment (alxndrkalinin) [inline app/scripts/analysis/charts/charts.coffee]: (style): all `->` should be surrounded by spaces

File: app/scripts/analysis/charts/charts.coffee [hunk 16]
```diff
@@ -1,86 +1,671 @@
...
+        .data(data)
+        .enter().append('circle')
+        .attr('class', 'dot')
+        .attr('r', 5)
+        .attr('cx', xMap)
+        .attr('cy', yMap)
+        .style('fill', (d)->color cValue(d))
+        .on('mouseover', (d)->
+          tooltip.transition().duration(200).style('opacity', .9)
+          tooltip.html('<div style="background-color:white; padding:5px; border-radius: 5px">(' + xValue(d)+ ',' + yValue(d) + ')</div>')
```
Comment (alxndrkalinin) [inline app/scripts/analysis/charts/charts.coffee]: (style): styles for elements should be defined in .less style file, e.g. see [styles for footer](https://github.com/SOCR/SOCRAT/blob/dev/app/styles/_general.less)

File: app/scripts/analysis/charts/charts.coffee (hunk 5 above)
Comment (nponarul) [inline app/scripts/analysis/charts/charts.coffee]: Done

File: app/scripts/analysis/charts/charts.coffee (hunk 8 above)
Comment (nponarul) [inline app/scripts/analysis/charts/charts.coffee]: Done

File: app/scripts/analysis/charts/charts.coffee (hunk 9 above)
Comment (nponarul) [inline app/scripts/analysis/charts/charts.coffee]: Done

File: app/scripts/analysis/charts/charts.coffee (hunk 10 above)
Comment (nponarul) [inline app/scripts/analysis/charts/charts.coffee]: Done

File: app/partials/analysis/charts/sidebar.jade (hunk 2 above)
Comment (nponarul) [inline app/partials/analysis/charts/sidebar.jade]: Done

File: app/scripts/analysis/charts/charts.coffee (hunk 4 above)
Comment (nponarul) [inline app/scripts/analysis/charts/charts.coffee]: Done

File: app/scripts/analysis/charts/charts.coffee (hunk 6 above)
Comment (nponarul) [inline app/scripts/analysis/charts/charts.coffee]: Done

File: app/scripts/analysis/charts/charts.coffee (hunk 7 above)
Comment (nponarul) [inline app/scripts/analysis/charts/charts.coffee]: Done

File: app/scripts/analysis/charts/charts.coffee (hunk 8 above)
Comment (nponarul) [inline app/scripts/analysis/charts/charts.coffee]: Done

File: app/scripts/analysis/charts/charts.coffee (hunk 9 above)
Comment (nponarul) [inline app/scripts/analysis/charts/charts.coffee]: Done

File: app/scripts/analysis/charts/charts.coffee (hunk 10 above)
Comment (nponarul) [inline app/scripts/analysis/charts/charts.coffee]: Done

File: app/scripts/analysis/charts/charts.coffee (hunk 11 above)
Comment (nponarul) [inline app/scripts/analysis/charts/charts.coffee]: Done

File: app/scripts/analysis/charts/charts.coffee (hunk 15 above)
Comment (nponarul) [inline app/scripts/analysis/charts/charts.coffee]: Done

**Pull request closed without merging**